### PR TITLE
MAID-3282: add malice protection for Requesting type

### PR DESCRIPTION
**Pull request by Fraser999**
This PR adds validation that incoming messages are correct (wrt the sender correctly following the `Requesting` → `Request` → `Response` flow) and also similar validation for all sync events as they're added to the graph.

In the case of the former, the message isn't handled at all (no events are added to our graph) and an accusation of `UnprovableMalice` is made against the sender.

To allow efficient checking that a sync event's corresponding descendant hasn't yet been added, a hash-set containing all the `Requesting` and `Request` sync events which don't yet have their partner `Request` or `Response` sync event has been added to `Graph`.

This also required the `dev_utils::Record` to be updated for the case where, at the end of iterating through the parsed dot file creating actions, we send the final few events via a new Request.  For that message to be valid, we now also create a `Requesting` sync event by the sender.  This means that after running `Record::play()`, the parsec instance will have two events not contained in the input dot file.

### CODE REVIEW

File: src/gossip/graph/mod.rs [hunk 1]
```diff
@@ -155,12 +177,116 @@ impl<P: PublicId> Graph<P> {
...
+                #[cfg(any(test, feature = "testing"))]
+                assert!(other_parent.is_requesting());
+                other_parent.index
+            });
+            let _removed =
+                index_to_remove.map(|index| self.awaiting_associated_events.remove(&index));
+            #[cfg(any(test, feature = "testing"))]
+            assert_eq!(_removed, Some(true));
+        } else if event.is_response() {
+            let index_to_remove = self.associated_request(event).map(|associated_request| {
```
Comment (jeanphilippeD) [inline src/gossip/graph/mod.rs]: why do you need associated_request instead of the other_parent index?

I do not understand, this function seem significantly more complicated than I would expect, what am i missing?

If adding to the graph a request or requesting, then add it to the hashset.
If adding to the graph a response or request, then remove the other parent hashset.

What am i missing?

File: src/gossip/graph/mod.rs [hunk 2]
```diff
@@ -155,12 +177,116 @@ impl<P: PublicId> Graph<P> {
...
+
 #[cfg(any(all(test, feature = "mock"), feature = "testing"))]
 impl<P: PublicId> Graph<P> {
     /// Remove the topologically last event.
     pub fn remove_last(&mut self) -> Option<(EventIndex, Event<P>)> {
         let event = self.events.pop()?;
         let _ = self.indices.remove(event.hash());
+        #[cfg(feature = "malice-detection")]
+        {
+            let _ = self
```
Comment (jeanphilippeD) [inline src/gossip/graph/mod.rs]: This only work if no malice, no?
Also, if awaiting_associated_events is a struct, then most should be implemented there.

File: src/gossip/messages.rs [hunk 3]
```diff
@@ -24,8 +34,42 @@ impl<T: NetworkEvent, P: PublicId> Request<T, P> {
...
 
-    pub(crate) fn hash_of_last_event_created_by(&self, src: &P) -> Result<Option<EventHash>> {
-        hash_of_last_event_created_by(src, &self.packed_events)
+    /// Returns `Ok` if the final event is a `Requesting` by `sender` targeting us, and for which we
+    /// have not yet recorded an associated `Request` event.  Otherwise returns `Err`.
+    #[cfg(feature = "malice-detection")]
+    pub(crate) fn validate<S: SecretId<PublicId = P>>(
+        &self,
+        sender: &P,
+        ctx: EventContextRef<T, S>,
```
Comment (jeanphilippeD) [inline src/gossip/messages.rs]: That we need to pass the context suggest this is not the right place for that function.
Would it be better in parsec_helper or new file?

File: src/gossip/messages.rs [hunk 4]
```diff
@@ -41,24 +85,64 @@ impl<T: NetworkEvent, P: PublicId> Response<T, P> {
...
+    ///   * for which we have not yet recorded an associated `Response` event
+    ///   * or if such an event immediately precedes a series of accusations by `sender`.
+    /// Otherwise returns `Err`.
+    #[cfg(feature = "malice-detection")]
+    pub(crate) fn validate<S: SecretId<PublicId = P>>(
+        &self,
+        sender: &P,
+        ctx: EventContextRef<T, S>,
+    ) -> Result<()> {
+        for packed_event in self.packed_events.iter().rev() {
```
Comment (jeanphilippeD) [inline src/gossip/messages.rs]: It feels somewhat more complicated than I would hope.
it seems it would be simpler to do as an in process thing since we are not trying to fight spam (i.e we will have processed all valid graph part and place them in the graph, but removed the invalid request/reponse and its descendent.

If it is done inside the unpack_and_add_events, as we are processing event, we can check each request other parent is a requesting with the correct destination, and each response other parent is a request with other parent having same creator as the response, and bail like we do for some of our malice detection.

File: src/dev_utils/record.rs [hunk 5]
```diff
@@ -162,13 +128,19 @@ impl From<ParsedContents> for Record {
                         "Sync event without other-parent: {:?}",
                         *event
                     );
+                    if event.is_request() {
+                        assert!(other_parent.is_requesting());
```
Comment (maqi) [inline src/dev_utils/record.rs]: better put some msg in case of assertion failure?
same comment to the other assertion below.

File: src/dev_utils/record.rs [hunk 6]
```diff
@@ -197,13 +169,15 @@ impl From<ParsedContents> for Record {
...
 
-        // Need a request to our peer with the remaining events
-        let (events_to_gossip, other_parent) = collect_remaining_event_to_gossip(&mut known);
-        if let Some(other_parent) = other_parent {
-            let src = unwrap!(contents
-                .peer_list
-                .get(other_parent.creator())
-                .map(|peer| peer.id().clone()));
+        let mut events_to_gossip = collect_remaining_events_to_gossip(&contents, &mut known);
+        let added_final_requesting_event = !events_to_gossip.is_empty();
```
Comment (maqi) [inline src/dev_utils/record.rs]: looking at this implementation, it seems the comment (`True if when parsing the graph we had`) of `added_final_requesting_event` and the name is bit confusing.
Do you mean whenever we parse the graph, we shall `add a final Requesting sync event and schedule a Request ` ?

File: src/dev_utils/record.rs [hunk 7]
```diff
@@ -197,13 +169,15 @@ impl From<ParsedContents> for Record {
...
+        if let Some(packed_event) = events_to_gossip.last() {
+            // Need a request to our peer with the remaining events
+            let src = packed_event.creator().clone();
```
Comment (maqi) [inline src/dev_utils/record.rs]: will there be chance that the last event in the `events_to_gossip` is not this peer?
in that case, `src` will be wrong?

File: src/gossip/graph/mod.rs (hunk 1 above)
Comment (Fraser999) [reply src/gossip/graph/mod.rs]: A lot of the complexity comes from handling response messages - more so than requests.

We need to ensure that the events in a response message should finish with a `Request` sync event, created by message sender, and that needs to have a `Requesting` as its other parent which is created by us and specifies the response sender as the recipient inside the `Requesting` event.

That's kinda complex, but what adds more is that a valid response message can have such a `Request` sync event not quite as the final one; that could validly be followed by a series of accusations all made by the sender.

In that case, when we receive that message and create our `Response` event, that will have the final accusation by the sender as its `other_parent`.  Hence the need to use `associated_request` here rather than just `other_parent`.

File: src/gossip/graph/mod.rs [hunk 8]
```diff
@@ -24,13 +26,19 @@ use std::collections::BTreeSet;
 pub(crate) struct Graph<P: PublicId> {
     events: Vec<Event<P>>,
     indices: BTreeMap<EventHash, EventIndex>,
+    /// Indices of `Requesting` events with no associated descendant `Request`, and `Request`s with
+    /// no associated descendant `Response`.
+    #[cfg(feature = "malice-detection")]
+    awaiting_associated_events: FnvHashSet<EventIndex>,
```
[the remark this answers is not in the document]
Comment (Fraser999) [reply src/gossip/graph/mod.rs]: I did consider that, but I don't think it would be worthwhile.  The bulk of the complexity around usage of `awaiting_associated_events` is in deducing what value should be passed as `other_parent` in your `insert()` method, and that's work which _should_ be done here in graph.rs rather than in a new struct.

If you disagree, I can try implementing this and see how it goes.

File: src/gossip/graph/mod.rs (hunk 2 above)
Comment (Fraser999) [reply src/gossip/graph/mod.rs]: Good point, I hadn't considered that last event being a malicious sync event.

File: src/gossip/messages.rs (hunk 3 above)
Comment (Fraser999) [reply src/gossip/messages.rs]: The only other place I'd see this belonging is in parsec.rs really.  I'd be fine to move the two `validate()` methods there if you prefer?  I don't like moving it to parsec_helper or a new file since not only would that still need the context passed, but also the message.

An advantage to keeping it here is that we can destructure `packed_event.content.cause`.  If I move it to parsec, I think I'd need to add one or two getters to `PackedEvent` similar to those on `Event` like `requesting_recipient()`.

File: src/functional_tests.rs [hunk 9]
```diff
@@ -241,8 +243,8 @@ fn add_peer() {
     // Generated with RNG seed: [411278735, 3293288956, 208850454, 2872654992].
     let mut parsed_contents = parse_test_dot_file("alice.dot");
 
-    // The final decision to add Fred is reached in D_18, so pop this event for now.
-    let d_18 = unwrap!(parsed_contents.remove_last_event());
+    // The final decision to add Fred is reached in E_25, so pop this event for now.
+    let e_25 = unwrap!(parsed_contents.remove_last_event());
```
Comment (maqi) [inline src/functional_tests.rs]: this change is due to the change in the req-resp pattern, or something else?

File: src/gossip/graph/mod.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (jeanphilippeD) [reply src/gossip/graph/mod.rs]: I agree, and think it is worthwhile.
I think while the object will be simple, it will simplify the complex code by removing some of the moving pieces.

File: src/gossip/messages.rs (hunk 4 above)
Comment (Fraser999) [reply src/gossip/messages.rs]: The issue is that we'd need to keep more state to track the events as we process them, or somehow flag the event which we think is the one which will decide whether we create our own sync event on the back of it.

Say I send a malicious response message which as its last packed events contains my malicious `Request` event, then a valid accusation made by me.  When you're processing the `Request`, you don't yet know if that final event (the accusation) is a valid `Request`, and that would change whether or not we want to create our own `Response` sync event.

Comment (jeanphilippeD) [reply src/gossip/messages.rs]: Not sure.
We do something similar for .

```
    fn add_event(&mut self, event: Event<S::PublicId>) -> Result<EventIndex> {
        let our = event.creator() == PeerIndex::OUR;
        if !our {
            #[cfg(feature = "malice-detection")]
            self.detect_malice(&event)?;
        }

...
// for example that one will stop the event from being added.
self.detect_other_parent_by_same_creator(event)?;
```


What I mean, is if we find any event that does not match the correct requesting/request/response, then we get an error, and stop adding any events left in the gossip, like we do for detect_other_parent_by_same_creator. This will also happen when creating our own sync event, i.e if the gossip was not correct, we will keep it, but not link it to a response sync event for example.

File: src/gossip/graph/mod.rs [hunk 10]
```diff
@@ -68,6 +85,11 @@ impl<P: PublicId> Graph<P> {
             }
         };
 
+        if needs_updated {
+            #[cfg(feature = "malice-detection")]
```
Comment (maqi) [inline src/gossip/graph/mod.rs]: why not put the feature guard outside the if clause?

File: src/gossip/graph/mod.rs [hunk 11]
```diff
@@ -155,12 +177,116 @@ impl<P: PublicId> Graph<P> {
...
+                sync_parent.is_request() && self.is_awaiting_associated_event(sync_parent.index)
+            })
+    }
+
+    fn update_awaiting(&mut self, index: EventIndex) {
+        let event = &self.events[index.0];
+        if event.is_requesting() {
+            let _inserted = self.awaiting_associated_events.insert(index);
+            #[cfg(any(test, feature = "testing"))]
+            assert!(_inserted);
```
Comment (maqi) [inline src/gossip/graph/mod.rs]: seems we expect the above `self.awaiting_associated_events.insert` always return true.
In that case, shall we put a log_or_panic ?
same comment to the below similar assertion places.

File: src/gossip/graph/mod.rs [hunk 12]
```diff
@@ -68,6 +85,11 @@ impl<P: PublicId> Graph<P> {
             }
         };
 
+        if needs_updated {
```
Comment (jeanphilippeD) [inline src/gossip/graph/mod.rs]: Why do we need this needs_updated at all?
What breaks if we call update_awaiting all the time?

Should we avoid updating if the event was already there instead?
It seems this function can be called multiple times.

We do not want to re-insert the requesting, if request has been found, afterward, but then we call that function with the requesting event again.

So it seems it should be done within the Vacant code.

File: src/gossip/graph/mod.rs (hunk 11 above)
Comment (jeanphilippeD) [inline src/gossip/graph/mod.rs]: do the asserts in this function reflect a bug, or an invalid input that we detect for malice? (if the latter, they probably should not be there).

File: src/gossip/graph/mod.rs [hunk 13]
```diff
@@ -155,12 +177,116 @@ impl<P: PublicId> Graph<P> {
...
+            let _inserted = self.awaiting_associated_events.insert(index);
+            #[cfg(any(test, feature = "testing"))]
+            assert!(_inserted);
+        } else if event.is_request() {
+            let _inserted = self.awaiting_associated_events.insert(index);
+            #[cfg(any(test, feature = "testing"))]
+            assert!(_inserted);
+            let index_to_remove = self.other_parent(event).map(|other_parent| {
+                #[cfg(any(test, feature = "testing"))]
+                assert!(other_parent.is_requesting());
```
Comment (jeanphilippeD) [inline src/gossip/graph/mod.rs]: I'm not sure this assert should be here, is that not more of a malice detection?

File: src/gossip/messages.rs (hunk 4 above)
Comment (Fraser999) [reply src/gossip/messages.rs]: That wouldn't solve the example I mentioned though.

File: src/gossip/graph/mod.rs [hunk 14]
```diff
@@ -155,12 +189,75 @@ impl<P: PublicId> Graph<P> {
...
+                .map(|associated_request| {
+                    #[cfg(any(test, feature = "testing"))]
+                    assert!(associated_request.is_request());
+                    associated_request.index
+                });
+            let _removed =
+                index_to_remove.map(|index| self.awaiting_associated_events.remove(&index));
+            #[cfg(any(test, feature = "testing"))]
+            assert_eq!(_removed, Some(true));
+        }
```
Comment (jeanphilippeD) [inline src/gossip/graph/mod.rs]: So I was thinking something along these lines:

```
fn get_awaiting_and_awaited_index(&self, index: EventIndex) {
    let event = &self.events[index.0];
    if event.is_requesting() {
        (Some(index), None)
    } else if event.is_request() {
        (Some(index), self.other_parent(event))
    } else if event.is_response() {
        (None, self
                .other_parent(event)
                .and_then(|other_parent| self.self_sync_ancestor(other_parent)))
    }
}

fn update_awaiting(&mut self, index: EventIndex) { 
    let (awaiting, awaited) = get_awaiting_and_awaited_index(index);
    if let Some(awaiting) = awaiting {
         let _inserted = self.awaiting_associated_events.insert(index);
    }
    if let Some(awaited) = awaited {
        let _removed = self.awaiting_associated_events.remove(awaited);
    }
}


// Or with the new struct, something like:
fn update_awaiting(&mut self, index: EventIndex) { 
    let (awaiting, awaited) = get_awaiting_and_awaited_index(index);
    let (_inserted, _removed) = self.awaiting_associated_events.insert(awaiting, awaited);
}
```

File: src/gossip/graph/mod.rs [hunk 15]
```diff
@@ -155,12 +189,75 @@ impl<P: PublicId> Graph<P> {
...
     /// Remove the topologically last event.
     pub fn remove_last(&mut self) -> Option<(EventIndex, Event<P>)> {
         let event = self.events.pop()?;
         let _ = self.indices.remove(event.hash());
+        #[cfg(feature = "malice-detection")]
+        {
+            let _ = self
+                .awaiting_associated_events
+                .remove(&EventIndex(self.events.len()));
+            if event.is_request() {
```
Comment (jeanphilippeD) [inline src/gossip/graph/mod.rs]: could you reuse the proposed get_awaiting_and_awaited_index here?

File: src/dev_utils/record.rs (hunk 5 above)
Comment (Fraser999) [reply src/dev_utils/record.rs]: I don't think I meant to commit this block.  I'll remove it since the second assertion is invalid in the case of accusations being made.  If you'd prefer me to keep it, let me know what additional info you'd want in the assertion and I'll add it.

File: src/dev_utils/record.rs (hunk 6 above)
Comment (Fraser999) [reply src/dev_utils/record.rs]: Yes, that's what the comment means :)
The comment was just moved from the previous location, but I agree it's probably less than helpful now.  I commented the new member `added_final_requesting_event` in the definition of `Record`.  Do you think that's enough so that this comment can just be removed?  Or should I more or less duplicate the comment from there?

File: src/dev_utils/record.rs (hunk 7 above)
Comment (Fraser999) [reply src/dev_utils/record.rs]: You mean so we'd be sending a `Request` message from ourself to ourself?  I don't think that's possible since all of the events we created have been turned into corresponding actions for us to do during `play()`.

I can add an assertion that `src != contents.our_id` if you want?

File: src/dev_utils/record.rs [hunk 16]
```diff
@@ -257,66 +232,128 @@ fn extract_genesis_group<'a>(
...
+            if thread::panicking() {
+                let msg = format!("!  Record constructed from '{}'  !", self.0.display());
+                let border = iter::repeat('!').take(msg.len()).collect::<String>();
+                println!("\n{1}\n{}\n{1}\n", msg, border);
             }
         }
     }
 
     /// Run smoke test using given dot file.
-    /// Use ignore_last_events to skip fake request that replay generates to send remaining gossip events.
```
[the remark this answers is not in the document]
Comment (Fraser999) [reply src/dev_utils/record.rs]: I did that when I defined `ignore_last_events` just below here.  I don't think it should still be a doc comment since it doesn't really apply to the signature/args of the function any more.

File: src/dev_utils/record.rs [hunk 17]
```diff
@@ -325,26 +362,37 @@ mod tests {
...
     #[test]
     fn smoke_parsec() {
-        let ignore_last_events = 0;
-        smoke("input_graphs/benches/minimal.dot", ignore_last_events)
+        let is_dot_file = |path: &PathBuf| {
+            path.extension()
+                .map(|extension| extension.to_string_lossy() == "dot")
+                .unwrap_or(false)
+        };
+        let max_file_size = 50_000;
```
[the remark this answers is not in the document]
Comment (Fraser999) [reply src/dev_utils/record.rs]: It yielded a decent number of extra dot files to get tested without slowing down this test much.  I don't really know if it's worthwhile adding that though as a comment, since it just changes the question from "Why 50,000?" to "How many extra dot files is enough?" and "What's an acceptable slowdown?"  I have added a check that we at least tested one dot file, in case our syntax changes again and all dot files end up over 50kB.

File: src/functional_tests.rs (hunk 9 above)
Comment (Fraser999) [reply src/functional_tests.rs]: This test was failing while I was still developing the code, and during debugging I noticed the name mismatch.  I don't know which previous commit caused the mismatch.

File: src/gossip/graph/mod.rs (hunk 12 above)
Comment (Fraser999) [reply src/gossip/graph/mod.rs]: > Why do we need this needs_updated at all?
What breaks if we call update_awaiting all the time?

I designed `update_awaiting()` to be called after we'd confirmed that this event wasn't already in our graph.  I _could_ change that behaviour so that `update_awaiting()` can be called unconditionally, but I don't think that's an improvement.

> Should we avoid updating if the event was already there instead?
It seems this function can be called multiple times.

I'm not quite following this.  You mean if the event was already in the graph, or in `awaiting_associated_events`?

> We do not want to re-insert the requesting, if request has been found, afterward, but then we call that function with the requesting event again.

I believe that's what `needs_updated` is doing for us here?

File: src/gossip/graph/mod.rs (hunk 10 above)
Comment (Fraser999) [reply src/gossip/graph/mod.rs]: I don't think that works for stable Rust.  I'd need to nest the `if` in a new scope to make that compile afaik.

File: src/gossip/graph/mod.rs (hunk 11 above)
Comment (Fraser999) [reply src/gossip/graph/mod.rs]: You mean change this to something like:
```rust
if !self.awaiting_associated_events.insert(index) {
    log_or_panic!("");
}
```
I could I guess - I just don't see the advantage?  Also, I'm not sure all these assertions will last for too long once we start writing tests for malicious behaviour

Comment (Fraser999) [reply src/gossip/graph/mod.rs]: They would reflect a bug; I was using them to help sanity check that `awaiting_associated_events` was being synced correctly in the absence of proper unit tests.  However, I'll remove them since they've served their purpose and depending on how we want to handle malicious entries, may be invalid anyway.

File: src/gossip/graph/mod.rs (hunk 13 above)
Comment (Fraser999) [reply src/gossip/graph/mod.rs]: I'll remove it.

File: src/dev_utils/record.rs (hunk 5 above)
Comment (maqi) [reply src/dev_utils/record.rs]: I am ok with you to remove this block. :)

File: src/dev_utils/record.rs (hunk 6 above)
Comment (maqi) [reply src/dev_utils/record.rs]: I think it's ok just to remove this comment.

File: src/dev_utils/record.rs (hunk 7 above)
Comment (maqi) [reply src/dev_utils/record.rs]: I thought the `events_to_gossip` contains the events in our graph that need to be sent to a peer, where some of those events are created by other peers (received not created).
Meanwhile, the `src` here needs to be `self`? which means   `events_to_gossip.last()` has to be an event created by self, which doesn't being reflected in the collection code?

File: src/dev_utils/record.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (maqi) [reply src/dev_utils/record.rs]: Ok then. It's just mandatory number is always suspicious.  :)

File: src/functional_tests.rs (hunk 9 above)
Comment (maqi) [reply src/functional_tests.rs]: maybe better split it into a separte commit as it's not realted to any of your change.

File: src/gossip/graph/mod.rs (hunk 11 above)
Comment (maqi) [reply src/gossip/graph/mod.rs]: if the assertion will be removed eventually, then I am ok with leave it as now.

File: src/dev_utils/record.rs (hunk 7 above)
Comment (Fraser999) [reply src/dev_utils/record.rs]: No, these are all incoming messages to us.

So, when constructing the `Record` we iterate the parsed dot file, and when we see e.g. a `Request` event created by us, it represents us having received a request message from the other_parent creator.  We don't prebuild any messages from ourself, we just call the running parsec's `create_gossip()` and `handle_request()` to allow it to generate the outgoing messages.

File: src/dev_utils/record.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (Fraser999) [reply src/dev_utils/record.rs]: Yeah, true.  I'll add a comment :)

File: src/functional_tests.rs (hunk 9 above)
Comment (Fraser999) [reply src/functional_tests.rs]: I _really_ don't want to have to do that for such a trivial change!  This is only renaming a variable.

File: src/gossip/graph/mod.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (Fraser999) [reply src/gossip/graph/mod.rs]: With your suggested change of `get_awaiting_and_awaited_index()` added now, do you agree it's not worth creating a wrapper for the set?

File: src/parsec.rs [hunk 18]
```diff
@@ -1721,6 +1713,65 @@ impl<T: NetworkEvent, S: SecretId> Parsec<T, S> {
...
 
+    fn detect_invalid_request(&mut self, event: &Event<S::PublicId>) -> Result<()> {
+        if !event.is_request() {
+            return Ok(());
+        }
+
+        let is_valid = self
+            .graph
+            .other_parent(event)
+            .map(|other_parent| {
```
Comment (jeanphilippeD) [inline src/parsec.rs]: I think I would call the closure parameter `requesting_event` here to match the other function.
Within the closure we care what event it represent, not where it is from.

File: src/parsec.rs [hunk 19]
```diff
@@ -1721,6 +1713,65 @@ impl<T: NetworkEvent, S: SecretId> Parsec<T, S> {
...
+            .unwrap_or(false);
+
+        if is_valid {
+            return Ok(());
+        }
+
+        let packed_event = event.pack(self.event_context())?;
+        self.create_accusation_event(
+            event.creator(),
+            Malice::InvalidRequest(Box::new(packed_event)),
```
Comment (jeanphilippeD) [inline src/parsec.rs]: any reason to have 2 types `InvalidRequest` and `InvalidResponse` instead of one maybe `InvalidRequestResponse`?

We could possibly even merge `detect_invalid_response` and `detect_invalid_request` with only the `is_valid` computation being different.

File: src/parsec.rs [hunk 20]
```diff
@@ -1721,6 +1713,65 @@ impl<T: NetworkEvent, S: SecretId> Parsec<T, S> {
...
+
+    fn detect_invalid_response(&mut self, event: &Event<S::PublicId>) -> Result<()> {
+        if !event.is_response() {
+            return Ok(());
+        }
+
+        let is_valid = self
+            .graph
+            .other_parent(event)
+            .and_then(|other_parent| self.graph.self_sync_ancestor(other_parent))
```
Comment (jeanphilippeD) [inline src/parsec.rs]: I think this is fine, (this may skip other events that are not accusation, which would be wrong would it not?)

I'm thinking we will probably want to move that check in unpack in the future when it is just checking other parent, like we do if we do not find any parent (i.e in unpack we probably want to ensure Requesting/Request/Response pattern, but not necessarily check that it is from correct. (maybe event only check that other_parent of an event is always Requesting/Request/Response but not check further there)).

File: src/parsec.rs [hunk 21]
```diff
@@ -1721,6 +1713,65 @@ impl<T: NetworkEvent, S: SecretId> Parsec<T, S> {
...
+            .and_then(|request_event| {
+                self.graph
+                    .other_parent(request_event)
+                    .map(|requesting_event| (request_event, requesting_event))
+            })
+            .map(|(request_event, requesting_event)| {
+                Some(request_event.creator()) == requesting_event.requesting_recipient()
```
Comment (jeanphilippeD) [inline src/parsec.rs]: Why do we check this? it seem it should have been validated by `detect_invalid_request` when request was added.

File: src/parsec.rs [hunk 22]
```diff
@@ -1450,19 +1454,12 @@ impl<T: NetworkEvent, S: SecretId> Parsec<T, S> {
...
     // of `is_request`) from `src`, then add it to our graph.
-    //
-    // `opt_other_parent` will contain the other-parent this new sync event should use, unless the
-    // gossip message from the peer was empty, in which case this will be `None` and we'll just use
-    // `src`'s most recent event we know of.
     fn create_sync_event(
         &mut self,
-        src_index: PeerIndex,
         is_request: bool,
+        other_parent: EventIndex,
```
Comment (jeanphilippeD) [inline src/parsec.rs]: I think this is a good bug fix.
Ideally it would be in its own commit. did that change an existing test that would cover it? should this bug fix have its own test?

File: src/parsec.rs [hunk 23]
```diff
@@ -1732,6 +1713,65 @@ impl<T: NetworkEvent, S: SecretId> Parsec<T, S> {
...
+    fn detect_invalid_request(&mut self, event: &Event<S::PublicId>) -> Result<()> {
+        if !event.is_request() {
+            return Ok(());
+        }
+
+        let is_valid = self
+            .graph
+            .other_parent(event)
+            .map(|other_parent| {
+                Some(event.creator()) == other_parent.requesting_recipient()
```
Comment (jeanphilippeD) [inline src/parsec.rs]: should we check is_requesting()?

File: src/parsec.rs [hunk 24]
```diff
@@ -1732,6 +1713,65 @@ impl<T: NetworkEvent, S: SecretId> Parsec<T, S> {
...
+            .graph
+            .other_parent(event)
+            .and_then(|other_parent| self.graph.self_sync_ancestor(other_parent))
+            .and_then(|request_event| {
+                self.graph
+                    .other_parent(request_event)
+                    .map(|requesting_event| (request_event, requesting_event))
+            })
+            .map(|(request_event, requesting_event)| {
+                Some(request_event.creator()) == requesting_event.requesting_recipient()
```
Comment (jeanphilippeD) [inline src/parsec.rs]: should we check is_request()?

File: src/parsec.rs (hunk 19 above)
Comment (Fraser999) [reply src/parsec.rs]: This was to match the current paradigm of being explicit in the `Malice` variants.  I'd rather not change this for now until we're sure there's no debugging benefit from keeping them separate.

File: src/parsec.rs (hunk 21 above)
Comment (Fraser999) [reply src/parsec.rs]: Yup, now that we don't add invalid `Request`s to the graph, this shouldn't ever return false.  Good catch.

File: src/parsec.rs (hunk 22 above)
Comment (Fraser999) [reply src/parsec.rs]: This isn't worthy of its own commit - it's fine as part of this.

There aren't any unit tests for these sorts of methods, but it's called in a few of the functional tests.

File: src/parsec.rs (hunk 23 above)
Comment (Fraser999) [reply src/parsec.rs]: That's implicit via `requesting_recipient()`, which will return `None` in that case.

File: src/parsec.rs (hunk 24 above)
Comment (Fraser999) [reply src/parsec.rs]: I don't think we'll need that once we nail down enforcing the Requesting→Request→Response pattern, but I'll add it for now.

File: src/parsec.rs [hunk 25]
```diff
@@ -1645,14 +1637,14 @@ impl<T: NetworkEvent, S: SecretId> Parsec<T, S> {
 
         self.detect_other_parent_by_same_creator(event)?;
         self.detect_self_parent_by_different_creator(event)?;
+        self.detect_invalid_request(event)?;
```
Comment (maqi) [inline src/parsec.rs]: at the begining of `detect_invalid_request & detect_invalid_response`, we will check `is_request or is_response`.
will there be a chance of this event is an invalid requesting ?

Comment (Fraser999) [reply src/parsec.rs]: I don't think so.  If the packed event was successfully unpacked, then we have a valid `Requesting` since (at least for now) there are no rules prohibiting creating it at any point.

Also, the calls to `is_request()` and `is_response()` are basically filtering all events except those from the corresponding `detect_...` methods - they're not really validating anything on their own.

File: src/parsec.rs (hunk 22 above)
Comment (jeanphilippeD) [reply src/parsec.rs]: What I meant, I guess, will a test break if we reverted that part?
If so, I'm happy as it is.

File: src/gossip/graph/mod.rs (hunk 12 above)
Comment (jeanphilippeD) [reply src/gossip/graph/mod.rs]: > I designed update_awaiting() to be called after we'd confirmed that this event wasn't already in our graph.

Comment (jeanphilippeD) [reply src/gossip/graph/mod.rs]: > I designed update_awaiting() to be called after we'd confirmed that this event wasn't already in our graph. 

> Should we avoid updating if the event was already there instead?
It seems this function can be called multiple times.

What I meant to say is this is exactly what the `match` is already doing, why do we have a second mechanism instead of using the `Entry::Vacant` branch.

File: src/parsec.rs (hunk 22 above)
Comment (Fraser999) [reply src/parsec.rs]: Yes, several tests fail if we use that old method of finding the other-parent.

File: src/gossip/graph/mod.rs (hunk 12 above)
Comment (Fraser999) [reply src/gossip/graph/mod.rs]: Can't remember what issue the borrow checker had before when I tried this, but with the recent changes, I can indeed just call `update_awaiting()` from inside the `Vacant` branch now.  Fixup commit pushed.
